index.js

```diff
@@ -348,6 +357,78 @@ function decorateCollection(collection) {
 
   };
 
+  newCollection.count = function(query, options, callback) {
+    if (arguments.length === 2) {
+      if ((typeof options) === 'function') {
+        callback = options;
+        options = {};
+      }
+    } else if (arguments.length === 1) {
+      if ((typeof query) === 'function') {
+        callback = query;
+        options = {};
+        query = {};
+      } else {
+        options = {};
+      }
+    } else if (!arguments.length) {
+      options = {};
+      query = {};
+    }
+    if (hasNestedProperties(query, [ '$where', '$near', '$nearSphere' ])) {
+      // Queries not supported by countDocuments must be turned into a
+      // find() that actually fetches the ids (minimum projection)
+      // and returns the number of documents
+      const cursor = collection.find(query);
+      if (options.limit !== undefined) {
+        cursor.limit(options.limit);
+      }
+      if (options.skip !== undefined) {
+        cursor.skip(options.skip);
+      }
+      if (options.hint !== undefined) {
+        cursor.hint(options.hint);
+      }
+      const p = cursor.project({ _id: 1 }).toArray().then(function(objects) {
+        if (callback) {
+          callback(null, objects.length);
```

Comment:
Why aren't we returning these callbacks and always returning `p`?

Reply:
p is a promise. Whether we want to return it depends on whether we were invoked with a callback.

Returning according to our normal "always be returning" pattern would return the promise *and* accept the callback, which leads to mongodb getting mad at us for "over-specifying."

This nonsense only comes up in rare situations like this where we support both promises and callbacks.